pkg/controller/nodeconfig/config/longhorn.go

```diff
@@ -0,0 +1,163 @@
+		if err := sys.TryRestartService(service); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func EnableV2DataEngine() error {
+	origHugepages, err := getNrHugepages()
+	if err != nil {
+		return err
+	}
+
+	// Write the persistent config first, so we know it's saved...
+	if err := UpdatePersistentOEMSettings(schema.Stage{
+		Name: spdkStageName,
+		Sysctl: map[string]string{
+			"vm.nr_hugepages": fmt.Sprintf("%d", hugepagesToAllocate),
+		},
+		Commands: []string{
+			"modprobe vfio_pci",
+			"modprobe uio_pci_generic",
+			"modprobe nvme_tcp",
+		},
+	}); err != nil {
+		return err
+	}
+
+	// ...then try to do the runtime activation (which may not succeed)
+	if err := modprobe(modulesToLoad, true); err != nil {
+		return fmt.Errorf("unable to load kernel modules %v: %v", modulesToLoad, err)
+	}
+
+	if origHugepages >= hugepagesToAllocate {
+		// We've already got enough hugepages, and don't want to unnecessarily
+		// restart the kubelet, so no further action required
+		return nil
+	}
+
+	if err := setNrHugepages(hugepagesToAllocate); err != nil {
+		return err
+	}
+
+	nrHugepages, err := getNrHugepages()
+	if err != nil {
+		return err
+	}
+	if nrHugepages == hugepagesToAllocate {
+		// We've successfully allocated the hugepages, but still need to restart
+		// the kubelet in order for Longhorn to see the allocation.
+		logrus.Infof("Restarting kubelet to set nr_hugepages=%d", hugepagesToAllocate)
+		return restartKubelet()
+	}
+
+	// We didn't get enough hugepages (not enough available unfragmented memory)
+	// but the system is now configured correctly so that if it's rebooted we should
+	// get the required allocation.
+	// TODO: record this somewhere (an event?) so that it can be picked up in the GUI
+	// Note that if there aren't enough hugepages, when harvester tries to enable the
+	// v2 data engine setting in Longhorn, the validator.longhorn.io admission webhook
+	// will pick up the failure and an error will be displayed on the harvester settings
+	// page, so we might not need to separately record this.
+	logrus.Errorf("Unable to allocate %d hugepages (only got %d)", hugepagesToAllocate, nrHugepages)
+
+	return nil
+}
+
+func DisableV2DataEngine() error {
+	origHugepages, err := getNrHugepages()
+	if err != nil {
+		return err
+	}
+
+	// Write the persistent config first, so we know it's saved...
+	if err := RemovePersistentOEMSettings(spdkStageName); err != nil {
+		return err
+	}
+
+	// ...then try to do the runtime deactivation
+	if err := modprobe(modulesToLoad, false); err != nil {
+		return fmt.Errorf("unable to unload kernel modules %v: %v", modulesToLoad, err)
+	}
+
+	if origHugepages == 0 {
+		// We already don't have any hugepages, and don't want to unnecessarily
+		// restart the kubelet, so no further action required
+		return nil
+	}
+
+	if err := setNrHugepages(0); err != nil {
+		return err
+	}
+
+	logrus.Info("Restarting kubelet to set nr_hugepages=0")
+	return restartKubelet()
```

Comment:
The very similar situation as above.
When we set the hugepage successfully but fail to restartKubelet, the controller will try to re-do it.
But the Hugepages is already the same as the target value, the restartKubelet() could not be called.

Reply:
Hrm..  I'm not sure how best to handle that case. TBH I'm inclined to ignore it. Just thinking through this... If `restartKubelet()` fails, my best guess is that the kubelet would be left in a non-running state. Then, if we called that function again, it wouldn't do anything, because it calls `TryRestartService` which only restarts services that are already running. So it probably needs user intervention at that point to figure out why the kubelet couldn't be restarted in the first place, and to fix whatever is wrong.

Reply:
ditto